### DISCUSSION TITLE
[scenario] Scaling servers with a local active server queue

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#25] [Q&A] [answered]
Let's say I start a [local active server queue](https://github.com/shikokuchuo/mirai#connecting-to-remote-servers-through-a-local-server-queue) with 2 nodes.

```r
daemons("tcp://192.168.0.2:5556", nodes = 2)
```

And then I start two servers, each on a different port in the contiguous block of ports, and each with a minute of max idle time. (By the way, does the idle time counter reset when a server receives a task?)

```r
# server 1
server("tcp://192.168.0.2:5556", idletime = 60000)
```

```r
# server 2
server("tcp://192.168.0.2:5557", idletime = 60000)
```

I submit a long task.

```r
m <- mirai(Sys.sleep(120))
```

This task goes to either server 1 or server 2, but I do not know which. By the second minute, one of the servers is still running the task, while the other times out and exits.

After that, a bunch of new tasks get queued, and I want to scale back up to 2 nodes. That means I need to submit a new call to `server()`. Should I use port 5556 or 5557 for that new server? How do I figure it out?

**Comment by shikokuchuo** [comment, marked as answer]
Honestly I think it will be easier with whatever you are using to send the job for execution on the server. Is there no channel there to see if the task has completed?

The alternative would be trying to peek into the background queue process and query active pipes. This is not impossible, but will add to the complexity, which I will prefer to avoid.

**Reply by shikokuchuo** [reply]
Or perhaps it is OK to just send a re-start to all servers - assuming that for the one that is still ongoing, it will just exit and restart once it's finished.

**Reply by wlandau** [reply]
> Honestly I think it will be easier with whatever you are using to send the job for execution on the server. Is there no channel there to see if the task has completed?

The server itself may keep running after the task is finished, so I was hoping to call `$data` on the `mirai` to see if the task is completed. (Is there a way to check status without downloading the result? Maybe with `$aio`?)

> The alternative would be trying to peek into the background queue process and query active pipes. This is not impossible, but will add to the complexity, which I will prefer to avoid.

Yeah, that does sound complicated and best to avoid.

> Or perhaps it is OK to just send a re-start to all servers - assuming that for the one that is still ongoing, it will just exit and restart once it's finished.

I am not sure I can picture how that would work with more general autoscaling, e.g. scaling back up to only half the original number of servers.

**Reply by shikokuchuo** [reply]
> > Honestly I think it will be easier with whatever you are using to send the job for execution on the server. Is there no channel there to see if the task has completed?
> 
> The server itself may keep running after the task is finished, so I was hoping to call `$data` on the `mirai` to see if the task is completed. (Is there a way to check status without downloading the result? Maybe with `$aio`?)
> 
Yes, it is simply `unresolved()` around the mirai object, which is TRUE only it it is still ongoing. This does technically pull in the data so R can see it (if it is there) - but it will be sitting in a buffer somewhere anyway as it has been sent (and received). It will never be the case that a task finishes and it is buffered at the sender.

**Reply by wlandau** [reply]
> The alternative would be trying to peek into the background queue process and query active pipes.

So knowing the TCP socket is not enough? Given only the string `"tcp:/40.1.1.20:5002"`, is it possible to to do a quick check from within the client process if there is a server dialing into client port 5002?

**Reply by shikokuchuo** [reply]
Not enough as the client will still be listening to that port regardless of whether there is a server dialled in. So you can't tell from the status of the port itself.

**Reply by shikokuchuo** [reply]
But thinking about this problem, I have thought up a solution with a high degree of cunning... to actually get the background queue process to report its connection status. Without it being cumbersome.

**Reply by shikokuchuo** [reply]
I have implemented this and it seems to work quite well :) Querying `daemons()` now returns the live status of the nodes. I will push this out once we've sorted your current issue.

**Reply by wlandau** [reply]
Amazing! I am eager to try it.

**Reply by wlandau** [reply]
I tried it, and the new `daemons()` that reports individual sockets is absolutely amazing! What a difference it makes!

**Reply by shikokuchuo** [reply]
It is in v0.7.2.9012 (https://github.com/shikokuchuo/mirai/commit/3f2d2d5d4a6c4cfa3c3c7458a776890c178b9032). No measurable impact on performance - that's the magic! In fact with small changes I've made over today / yesterday perfomance is actually better than before.

**Reply by wlandau** [reply]
Magic indeed!
